### PR TITLE
FFTW: enable threading, long-double, quad, optimalization

### PR DESCRIPTION
**Pull request by extrowerk**
Needs to be tested on gcc2h

### CODE REVIEW

**Comment by extrowerk** [comment]
Oh it doesn't even build on gcc2h because the fortran dep.
Then let's disable it for gcc2.

**Comment by waddlesplash** [comment]
Can we add a non-Fortran version for GCC2, then?

**Comment by extrowerk** [comment]
Can somebody test this pr on gcc2 (so do not add _x86 postfix)?

File: sci-libs/fftw/fftw-3.3.8.recipe [hunk 1]
```diff
@@ -63,33 +95,86 @@ BUILD_PREREQUIRES="
...
+	cd ..
 }
 
 INSTALL()
 {
+	if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
+		for f in double long-double quad float; do
+	else
+		for f in double long-double float; do
+	fi
```
Comment (fbrosson) [inline sci-libs/fftw/fftw-3.3.8.recipe]:
```
INSTALL()
{
	if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
		for f in double long-double quad float; do
	else
		for f in double long-double float; do
	fi
			cd $f
			make install
			cd ..
		done
```
This won't work. Unlike a `#if` + `#else` + `#endif`, the `if` + `else` + `fi`  can't include the `for` + `do` and let the `done` outside.

Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: The easiest way to address my previous comment is:
```diff
 INSTALL()
 {
 	if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
-		for f in double long-double quad float; do
+		DIRS="double long-double quad float"
 	else
-		for f in double long-double float; do
+		DIRS="double long-double float"
 	fi
+	for f in $DIRS; do
- 		cd $f
 		make -C $f install
- 		cd ..
 	done
 
- 	rm $libDir/libfftw*.la
+ 	rm -f "$libDir"/libfftw*.la 
```
(The very last change, for the `rm $libDir/libfftw*.la` is not important.)

And of course, the `make -C $f install` saves us the `cd $f` and the `cd ..` so I think it can be a nice idea.

Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Hmm, here's a more compact variant:
```diff
 INSTALL()
 {
 	if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
-		for f in double long-double quad float; do
+		for f in double long-double quad float; do make -C $f install; done
 	else
-		for f in double long-double float; do
+		for f in double long-double float; do make -C $f install; done
 	fi
- 		cd $f
- 		make install
- 		cd ..
- 	done
```

File: sci-libs/fftw/fftw-3.3.8.recipe [hunk 2]
```diff
@@ -63,33 +95,86 @@ BUILD_PREREQUIRES="
...
 defineDebugInfoPackage fftw$secondaryArchSuffix \
-	$libDir/libfftw3.so.$libVersion \
-	$libDir/libfftw3f.so.$libVersion
+	"$libDir"/libfftw3.so.$libVersion \
+	"$libDir"/libfftw3_threads.so.$libVersion \
+	"$libDir"/libfftw3f.so.$libVersion \
+	"$libDir"/libfftw3f_threads.so.$libVersion \
+	"$libDir"/libfftw3l.so.$libVersion \
+	"$libDir"/libfftw3l_threads.so.$libVersion \
+	"$libDir"/libfftw3q.so.$libVersion \
```
Comment (fbrosson) [inline sci-libs/fftw/fftw-3.3.8.recipe]:
```diff
-defineDebugInfoPackage fftw$secondaryArchSuffix \
+debugList=(
-	"$libDir"/libfftw3.so.$libVersion \
+	"$libDir"/libfftw3.so.$libVersion
-	"$libDir"/libfftw3_threads.so.$libVersion \
+	"$libDir"/libfftw3_threads.so.$libVersion
-	"$libDir"/libfftw3f.so.$libVersion \
+	"$libDir"/libfftw3f.so.$libVersion
-	"$libDir"/libfftw3f_threads.so.$libVersion \
+	"$libDir"/libfftw3f_threads.so.$libVersion
-	"$libDir"/libfftw3l.so.$libVersion \
+	"$libDir"/libfftw3l.so.$libVersion
-	"$libDir"/libfftw3l_threads.so.$libVersion \
+	"$libDir"/libfftw3l_threads.so.$libVersion
+	)
+if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
+debugList+=(
-	"$libDir"/libfftw3q.so.$libVersion \
+	"$libDir"/libfftw3q.so.$libVersion
 	"$libDir"/libfftw3q_threads.so.$libVersion
+	)
+fi
+defineDebugInfoPackage fftw$secondaryArchSuffix "${debugList[@]}"
```
This should work.
(I did not test it.)

File: sci-libs/fftw/fftw-3.3.8.recipe [hunk 3]
```diff
@@ -63,33 +95,86 @@ BUILD_PREREQUIRES="
...
+	cd ..
+
+	# build long-double-precision version
+	cd long-double
+	runConfigure ../configure \
+		--enable-long-double \
+		$CONFIGURE
 	make $jobArgs
-	make install
+	cd ..
```
Comment (fbrosson) [inline sci-libs/fftw/fftw-3.3.8.recipe]: Two things in this comment:
* A suggestion to use `pushd long-double` and `popd` instead of `cd long-double` and `cd ..`.
* Is it on purpose that you are not using `$CONFIGURE_X86` for `long-double`, although you are using it for `double`? Important: I don't know if using it for `long-double` would be appropriate. I"m just asking.

```diff
 	# build long-double-precision version
-	cd long-double
+	pushd long-double
 	runConfigure ../configure \
 		--enable-long-double \
- 		$CONFIGURE
+ 		$CONFIGURE \
+		$CONFIGURE_X86
 	make $jobArgs
-	cd ..
+	popd
```
BTW, if you like the `pushd` & `popd` then you can use them for the other blocs too.

Comment (extrowerk) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Only single and double can handle sse2.

File: sci-libs/fftw/fftw-3.3.8.recipe [hunk 4]
```diff
@@ -151,17 +158,15 @@ fi
...
-			cd $f
-			make install
-			cd ..
-		done
-
-	rm $libDir/libfftw*.la
+	DIRS="double long-double float"
+ 	if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
+		DIRS="$DIRS quad"
+ 	fi
```
[the remark this answers is not in the document]
Comment (extrowerk) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Thanks!

File: sci-libs/fftw/fftw-3.3.8.recipe (hunk 3 above)
Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Ok, thanks.

File: sci-libs/fftw/fftw-3.3.8.recipe (hunk 4 above)
[the remark this answers is not in the document]
Comment (Begasus) [reply sci-libs/fftw/fftw-3.3.8.recipe]: I think part of this could also be used in TEST()?

[the remark this answers is not in the document]
Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: I guess you are right. This would be nice btw.
So, if using `DIRS` in `TEST()` is desirable than you can move the definition of `DIRS` (and `fftwLibs`) above `INSTALL()`:

```diff

DIRS="double long-double float"
fftwLibs="libfftw3 libfftw3_threads libfftw3f libfftw3f_threads libfftw3l libfftw3l_threads"
if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
	DIRS="$DIRS quad"
	fftwLibs="$fftwLibs libfftw3q libfftw3q_threads"
fi

INSTALL()
{
	prepareInstalledDevelLibs $fftwLibs
 	fixPkgconfig
```

[the remark this answers is not in the document]
Comment (Begasus) [reply sci-libs/fftw/fftw-3.3.8.recipe]: That would be even better :)

**Comment by Begasus** [comment]
Builds and tests are ok for x86_gcc2 (confirmed)

File: sci-libs/fftw/fftw-3.3.8.recipe [hunk 5]
```diff
@@ -15,42 +15,74 @@ HOMEPAGE="http://www.fftw.org/"
...
 libVersionCompat="$libVersion compat >= ${libVersion%%.*}"
 
 PROVIDES="
 	fftw$secondaryArchSuffix = $portVersion compat >= 3
 	cmd:fftw_wisdom$secondaryArchSuffix
-	cmd:fftw_wisdom_to_conf$secondaryArchSuffix
 	cmd:fftwf_wisdom$secondaryArchSuffix
+	cmd:fftwl_wisdom$secondaryArchSuffix
+	cmd:fftwq_wisdom$secondaryArchSuffix
+	cmd:fftw_wisdom_to_conf$secondaryArchSuffix
```
Comment (alaviss) [inline sci-libs/fftw/fftw-3.3.8.recipe]: I think we should drop the suffix for x86_gcc2.

Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Do you mean: "drop `$secondaryArchSuffix` for the `cmd:` runtimes" ?

Since the recipe builds on all primary arches, what we _could_ do, instead, is drop some of the `cmd:` runtimes.
But I think the PR is better as it is right now.

Comment (alaviss) [reply sci-libs/fftw/fftw-3.3.8.recipe]: Oh, my bad, I didn't notice. I think we should drop those `cmd:` for secondary architectures, their functions doesn't appear to deviate between architecture to me.

Comment (fbrosson) [reply sci-libs/fftw/fftw-3.3.8.recipe]: I don't have a strong opinion on whether or not we should drop the `cmd:` runtimes for secindary arches.
I'd say that keeping them **_might_** help x86_gcc2h users who could choose to call the `-x86` versions, as these have access to the quad libs, `libfftw3q{,_threads}.so`, which are not available on x86_gcc2 primary arch.

So, we could choose to drop the `cmd:` runtimes only for all secondary arch builds on platforms x86_64 (and x86), and keep them for x86 secondary arch of x86_gcc2h.

So, to summarize, there are 3 scenari:
1) Keep all `cmd:` in all cases. (That's what we have right now in the PR.)
2) Keep the `cmd:` on secondary arch if `$targetArchitecture" = x86_gcc2.
3) Drop all `cmd:` on all secondary architectures.

I don't know which is best. All I know is that solution 1 covers all possible needs.

**Comment by fbrosson** [comment]
Some tests are failing on x86_64:
<Details>

```
Making check in tests
make[1]: Entering directory '/sources/fftw-3.3.8/long-double/tests'
make  check-local
make[2]: Entering directory '/sources/fftw-3.3.8/long-double/tests'
perl -w ../../tests/check.pl  -r -c=30 -v `pwd`/bench

...

obcd66v43 2.07618e-19 2.18496e-19 2.17441e-19
ibcd66v43 1.56479e-19 2.73119e-19 2.58186e-19
--------------------------------------------------------------
         FFTW transforms passed basic tests!
--------------------------------------------------------------
perl -w ../../tests/check.pl  -r -c=30 -v --nthreads=2 `pwd`/bench
Executing "/sources/fftw-3.3.8/long-double/tests/bench -o nthreads=2 --verbose=1   --verify 'ofc9x10x5*9' --verify 'ifc9x10x5*9' --verify 'okd11880o10' --verify 'ikd11880o10' --verify 'obr6x3x9*5' --verify 'ibr6x3x9*5' --verify 'ofr6x3x9*5' --verify 'ifr6x3x9*5' --verify '//obc6x3x9*5' --verify '//ibc6x3x9*5' --verify '//ofc6x3x9*5' --verify '//ifc6x3x9*5' --verify 'obc6x3x9*5' --verify 'ibc6x3x9*5' --verify 'ofc6x3x9*5' --verify 'ifc6x3x9*5' --verify 'ok13e10x13e10x2o10x5o11' --verify 'ik13e10x13e10x2o10x5o11' --verify 'obr16x36v1' --verify 'ibr16x36v1' --verify 'ofr16x36v1' --verify 'ifr16x36v1' --verify '//obc16x36v1' --verify '//ibc16x36v1' --verify '//ofc16x36v1' --verify '//ifc16x36v1' --verify 'obc16x36v1' --verify 'ibc16x36v1' --verify 'ofc16x36v1' --verify 'ifc16x36v1' --verify 'okd2304o00' --verify 'ikd2304o00' --verify 'obrd6x10*34' --verify 'ibrd6x10*34' --verify 'ofrd6x10*34' --verify 'ifrd6x10*34' --verify '//obcd6x10*34' --verify '//ibcd6x10*34' --verify '//ofcd6x10*34' --verify '//ifcd6x10*34' --verify 'obcd6x10*34' --verify 'ibcd6x10*34' --verify 'ofcd6x10*34' --verify 'ifcd6x10*34'"
ofc9x10x5*9 1.79777e-19 2.72585e-19 3.95886e-19
ifc9x10x5*9 2.22796e-19 2.72585e-19 3.44145e-19
bench: ../../libbench2/verify-r2r.c:61: assertion failed: !isnan(e)
FAILED /sources/fftw-3.3.8/long-double/tests/bench:  --verify 'ofc9x10x5*9' --verify 'ifc9x10x5*9' --verify 'okd11880o10' --verify 'ikd11880o10' --verify 'obr6x3x9*5' --verify 'ibr6x3x9*5' --verify 'ofr6x3x9*5' --verify 'ifr6x3x9*5' --verify '//obc6x3x9*5' --verify '//ibc6x3x9*5' --verify '//ofc6x3x9*5' --verify '//ifc6x3x9*5' --verify 'obc6x3x9*5' --verify 'ibc6x3x9*5' --verify 'ofc6x3x9*5' --verify 'ifc6x3x9*5' --verify 'ok13e10x13e10x2o10x5o11' --verify 'ik13e10x13e10x2o10x5o11' --verify 'obr16x36v1' --verify 'ibr16x36v1' --verify 'ofr16x36v1' --verify 'ifr16x36v1' --verify '//obc16x36v1' --verify '//ibc16x36v1' --verify '//ofc16x36v1' --verify '//ifc16x36v1' --verify 'obc16x36v1' --verify 'ibc16x36v1' --verify 'ofc16x36v1' --verify 'ifc16x36v1' --verify 'okd2304o00' --verify 'ikd2304o00' --verify 'obrd6x10*34' --verify 'ibrd6x10*34' --verify 'ofrd6x10*34' --verify 'ifrd6x10*34' --verify '//obcd6x10*34' --verify '//ibcd6x10*34' --verify '//ofcd6x10*34' --verify '//ifcd6x10*34' --verify 'obcd6x10*34' --verify 'ibcd6x10*34' --verify 'ofcd6x10*34' --verify 'ifcd6x10*34'
Makefile:686: recipe for target 'check-local' failed
make[2]: *** [check-local] Error 1
make[2]: Leaving directory '/sources/fftw-3.3.8/long-double/tests'
Makefile:561: recipe for target 'check-am' failed
make[1]: *** [check-am] Error 2
make[1]: Leaving directory '/sources/fftw-3.3.8/long-double/tests'
Makefile:708: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1
make: Leaving directory '/sources/fftw-3.3.8/long-double'
```

</Details>
